modular_skyrat/modules/marines/code/smartgun.dm

```diff
@@ -0,0 +1,167 @@
+		return
+	cover_open = !cover_open
+	to_chat(user, span_notice("You [cover_open ? "open" : "close"] [src]'s cover."))
+	playsound(src, 'sound/weapons/gun/l6/l6_door.ogg', 60, TRUE)
+	update_appearance()
+	return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
+
+/obj/item/gun/ballistic/automatic/smart_machine_gun/can_shoot()
+	if(cover_open)
+		balloon_alert_to_viewers("cover open!")
+		return FALSE
+	return chambered
+
+/obj/item/gun/ballistic/automatic/smart_machine_gun/eject_magazine(mob/user, display_message = TRUE, obj/item/ammo_box/magazine/tac_load = null)
+	if(!cover_open)
+		to_chat(user, span_warning("The cover is closed! Open it before ejecting the magazine!"))
+		return
+	return ..()
+
+/obj/item/gun/ballistic/automatic/smart_machine_gun/attackby(obj/item/attack_item, mob/user, params)
+	if(!cover_open && istype(attack_item, mag_type))
+		to_chat(user, span_warning("[src]'s dust cover prevents a magazine from being fit."))
+		return
+	..()
+
+/obj/item/gun/ballistic/automatic/smart_machine_gun/update_overlays()
+	. = ..()
+	. += "[base_icon_state]_door_open"
+
+/obj/item/gun/ballistic/automatic/smart_machine_gun/unrestricted
+    pin = /obj/item/firing_pin 
+// Magazine itself
+
+/obj/item/ammo_box/magazine/smartgun_drum
+	name = "smartgun drum (10x28mm caseless)"
+	icon = 'modular_skyrat/modules/marines/icons/items/ammo.dmi'
+	icon_state = "smartgun_drum"
+	ammo_type = /obj/item/ammo_casing/smart/caseless/a10x28
+	caliber = "a10x28"
+	max_ammo = 500
+	multiple_sprites = AMMO_BOX_FULL_EMPTY
+	w_class = WEIGHT_CLASS_NORMAL
+
+// Smart ammo casings
+
+/obj/item/ammo_casing/smart
+	icon = 'modular_skyrat/modules/gunsgalore/icons/ammo/ammo.dmi'
+
+/obj/item/ammo_casing/smart/Initialize(mapload)
+	. = ..()
+	RegisterSignal(src, COMSIG_CHAMBERED_BULLET_FIRE, .proc/iff_transfer)
+
+/obj/item/ammo_casing/smart/proc/iff_transfer(datum/source, list/iff_factions)
+	SIGNAL_HANDLER
+
+	if(istype(loaded_projectile, /obj/projectile/bullet/smart))
+		var/obj/projectile/bullet/smart/smart_proj = loaded_projectile
+		smart_proj.iff_factions = iff_factions.Copy()
+
+/obj/item/ammo_casing/smart/caseless
+	firing_effect_type = null
+	heavy_metal = FALSE
+
+/obj/item/ammo_casing/smart/caseless/fire_casing(atom/target, mob/living/user, params, distro, quiet, zone_override, spread, atom/fired_from)
+	if (!..()) //failed firing
+		return FALSE
+	if(istype(fired_from, /obj/item/gun))
+		var/obj/item/gun/shot_from = fired_from
+		if(shot_from.chambered == src)
+			shot_from.chambered = null //Nuke it. Nuke it now.
+	qdel(src)
+	return TRUE
+
+/obj/item/ammo_casing/smart/caseless/update_icon_state()
+	. = ..()
+	icon_state = "[initial(icon_state)]"
+
+/obj/item/ammo_casing/smart/caseless/a10x28
+	name = "10x28mm bullet"
+	desc = "A 10x28m caseless bullet."
+	icon_state = "792x57-casing"
+	caliber = "a10x28"
+	projectile_type = /obj/projectile/bullet/smart/a10x28
+
+// Smart bullets
+
+/obj/projectile/bullet/smart
+	var/list/iff_factions = list()
```

Comment:
a variable that works like this already exists

Reply:
Is there? I can't seem to find it.

Reply:
Will work on this tomorrow since i coded the smartgun